cpp/ql/src/experimental/Security/CWE/CWE-787/UnsignedToSignedPointerArith.ql

```diff
@@ -0,0 +1,28 @@
+/**
+ * @author Jordy Zomer
+ * @name unsiged to signed used in pointer arithmetic
+ * @description finds unsigned to signed conversions used in pointer arithmetic, potentially causing an out-of-bound access
+ * @id cpp/out-of-bounds
+ * @kind problem
+ * @problem.severity warning
+ * @tags reliability
+ *       security
+ *       external/cwe/cwe-787
+ */
+
+import cpp
+import semmle.code.cpp.dataflow.DataFlow
+import semmle.code.cpp.security.Overflow
+
+from FunctionCall call, Function f, Parameter p, DataFlow::Node sink, PointerArithmeticOperation pao, Operation a, Operation b
+where
+f = call.getTarget() and
+p = f.getAParameter() and
+p.getType().getUnderlyingType().(IntegralType).isSigned() and
+call.getArgument(p.getIndex()).getType().getUnderlyingType().(IntegralType).isUnsigned() and
```

Comment:
You want to use `getUnspecifiedType()` here, which will strip specifiers as well as typedefs. We represent `const` and `volatile` as wrappers, so `const unsigned int` will be a `SpecifiedType` that wraps the `IntegralType` instance for `unsigned int`, and it won't match here.

As a stylistic note, you can also drop the `getType()` calls and call the `getUnspecifiedType()` predicate directly on `p` and `call`.